compiler/rustc_parse/src/parser/ty.rs

```diff
@@ -360,13 +391,37 @@ impl<'a> Parser<'a> {
             }
             Err(err) => return Err(err),
         };
-        let ty = if self.eat(&token::Semi) {
-            TyKind::Array(elt_ty, self.parse_anon_const_expr()?)
-        } else {
-            TyKind::Slice(elt_ty)
-        };
-        self.expect(&token::CloseDelim(token::Bracket))?;
-        Ok(ty)
+        let size_const =
+            if self.eat(&token::Semi) { Some(self.parse_anon_const_expr()?) } else { None };
+
+        match self.expect(&token::CloseDelim(token::Bracket)) {
+            Ok(_) => Ok(match size_const {
+                Some(size_const) => TyKind::Array(elt_ty, size_const),
+                None => TyKind::Slice(elt_ty),
+            }),
+            Err(mut err1)
+                if recover_array_type == RecoverArrayType::Yes
```

Comment:
Is there _any_ case where we wouldn't want to do the lookahead you're doing here?

Reply:
I'm not doing lookahead here, do you mean recovery? See the `RecoverArrayType` documentation for why we don't want to try to recover always. Also documented in the commit message. In short, it makes some error messages in pre-2018 editions worse.

I think that may be fine as we probably don't have a lot of users today that actively write pre-2018 code (and I'd expect those people to use an older version of rustc, without this patch). If you agree that this is fine, then the `RecoverArrayType` type and extra parameter can be removed.